ledger/simulation/resources.go

```diff
@@ -0,0 +1,391 @@
+	var startingBoxes int
+	localTxnResources := make([]ResourceAssignment, len(txns))
+	for i := range txns {
+		localTxnResources[i] = makeTxnResourceAssignment(&txns[i].Txn, proto)
+		startingBoxes += len(txns[i].Txn.Boxes)
+	}
+	return GroupResourceAssignment{
+		Resources:         makeGlobalResourceAssignment(localTxnResources, proto),
+		localTxnResources: localTxnResources,
+		startingBoxes:     startingBoxes,
+	}
+}
+
+func (a *GroupResourceAssignment) removePrivateFields() {
+	a.startingBoxes = 0
+	a.localTxnResources = nil
+}
+
+func (a *GroupResourceAssignment) hasAccount(addr basics.Address, ep *logic.EvalParams, programVersion uint64, globalSharing bool, txnIndex int) bool {
+	if globalSharing {
+		return a.Resources.hasAccount(addr, ep, programVersion)
+	}
+	return a.localTxnResources[txnIndex].hasAccount(addr, ep, programVersion)
+}
+
+func (a *GroupResourceAssignment) addAccount(addr basics.Address, globalSharing bool, txnIndex int) bool {
+	if globalSharing {
+		return a.Resources.addAccount(addr)
+	}
+	return a.localTxnResources[txnIndex].addAccount(addr)
+}
+
+func (a *GroupResourceAssignment) hasAsset(aid basics.AssetIndex, globalSharing bool, txnIndex int) bool {
+	if globalSharing {
+		return a.Resources.hasAsset(aid)
+	}
+	return a.localTxnResources[txnIndex].hasAsset(aid)
+}
+
+func (a *GroupResourceAssignment) addAsset(aid basics.AssetIndex, globalSharing bool, txnIndex int) bool {
+	if globalSharing {
+		return a.Resources.addAsset(aid)
+	}
+	return a.localTxnResources[txnIndex].addAsset(aid)
+}
+
+func (a *GroupResourceAssignment) hasApp(aid basics.AppIndex, globalSharing bool, txnIndex int) bool {
+	if globalSharing {
+		return a.Resources.hasApp(aid)
+	}
+	return a.localTxnResources[txnIndex].hasApp(aid)
+}
+
+func (a *GroupResourceAssignment) addApp(aid basics.AppIndex, ep *logic.EvalParams, programVersion uint64, globalSharing bool, txnIndex int) bool {
+	if globalSharing {
+		return a.Resources.addApp(aid, ep, programVersion)
+	}
+	return a.localTxnResources[txnIndex].addApp(aid, ep, programVersion)
+}
+
+func (a *GroupResourceAssignment) hasBox(app basics.AppIndex, name string) bool {
+	// all boxes are global, never consult PerTxnResources
+	return a.Resources.hasBox(app, name)
+}
+
+func (a *GroupResourceAssignment) addBox(app basics.AppIndex, name string) bool {
+	// all boxes are global, never consult PerTxnResources
+	return a.Resources.addBox(app, name)
+}
+
+func (a *GroupResourceAssignment) hasHolding(addr basics.Address, aid basics.AssetIndex) bool {
+	// nil map lookup is ok
+	_, ok := a.AssetHoldings[ledgercore.AccountAsset{Address: addr, Asset: aid}]
+	return ok
+}
+
+func (a *GroupResourceAssignment) addHolding(addr basics.Address, aid basics.AssetIndex) bool {
+	// TODO: limit cross-product usage
+	if a.AssetHoldings == nil {
+		a.AssetHoldings = make(map[ledgercore.AccountAsset]struct{})
+	}
+	a.AssetHoldings[ledgercore.AccountAsset{Address: addr, Asset: aid}] = struct{}{}
```

Comment:
maybe a dumb question: if we are allowing for holding/local with both asset/app and accont available, do we want to first try add asset/app and account, then check cross product availability?

Reply:
This is a really good question, and this is the reason I added the extra individual checks for the account and asset here: https://github.com/algorand/go-algorand/blob/7de6599a945ca9d89071bd11bf50f64c2ed95b89/data/transactions/logic/resources.go#L186-L189

If we don't do this, then `AllowsHolding` does not know if the reason it was called was because one or both of the account or asset isn't available, or if they are both available and the holding is the thing that isn't available.

Note that `cx.availableAccount(addr)` and `cx.availableAsset(ai)` will each check if the resource is available normally, and if not they will ask `cx.UnnamedResources`. So both cases above are covered.